### PR TITLE
Improve results object when using SCIP

### PR DESCRIPTION
**Pull request by pmlpm1986**
## Fixes # .

Previously, the 'aborted' state would be used to characterise a variety of solver statuses (in the SCIPAMPL.py file) irrespective of their internal implications and meaning. This has been corrected.
The results object also lacked information about the optimisation process, such as the (MIP) gap, the actual (SCIP-provided) solving time, and the primal and dual bounds. This is now provided via the 'Solver' dict of the results object.

## Summary/Motivation:
There was previously no way to get thet (MIP) gap when solving problems through SCIP.

I also wanted more information, which the results object does not provide. Some of them are mentioned in: https://github.com/Pyomo/pyomo/issues/1615

## Changes proposed in this PR:
- Coherent solver statuses with SCIP
- More optimisation process data available to the user

## Other comments

This PR seeks to correct issues when using the SCIP solver, namely concerning the results object.

The results object now includes information about the optimisation process, as found in the log file. This includes the time needed for the optimisation (in seconds; as computed by SCIP itself, not Pyomo), the (MIP) gap (in %), the primal bound and the dual bound. This is now found under the 'Solver' key.

The SolverStatus and TerminationCondition attribute combinations have also been revised to avoid warnings. 

Other problems remain. The 'Problem' part of the results object is updated in ways that were not understood. As such, no changes were made to that dict, but it is logical to use the 'Solver' dict to update the lower and upper bounds. Overall, the interaction with solvers does not appear to be sufficiently standardised, let alone documented, making pyomo (way) less appealing than it should be. It does work, of course, but making changes to it is not for the faint of heart.

One final note about getting information from SCIP: it can be asked to generate  a statistics file, which contains all the information one is likely to need to fill the results object with data. Since that would involve messing with the SCIP command, I decided not to do it for the time being.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 1]
```diff
@@ -199,15 +199,76 @@ def create_command_line(self, executable, problem_files):
         return Bunch(cmd=cmd, log_file=self._log_file, env=env, cwd=options_dir)
 
     def _postsolve(self):
-        results = super(SCIPAMPL, self)._postsolve()
-        if results.solver.message == "unknown":
+        # results = super(SCIPAMPL, self)._postsolve() # previous way
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:202]: Commented out - can be removed

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 2]
```diff
@@ -199,15 +199,76 @@ def create_command_line(self, executable, problem_files):
...
+                if (not self._soln_file is None) and \
+                   os.path.exists(self._soln_file):
+                    os.remove(self._soln_file)
+
+        TempfileManager.pop(remove=not self._keepfiles)
+        
+        #**********************************************************************
+        
+        # # UNKNOWN
+        # unknown='unknown'                               # An unitialized value
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:258]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line? (Also, note, typo in this particular comment)
```
## UNKNOWN
# unknown=`unknown`
# An uninitialized value 
```

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 3]
```diff
@@ -199,15 +199,76 @@ def create_command_line(self, executable, problem_files):
...
             results.solver.status = \
                 SolverStatus.unknown
             results.solver.termination_condition = \
                 TerminationCondition.unknown
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.unknown
+                    
+        # # ABORTED
+        # userInterrupt='userInterrupt'                   # Interrupt signal generated by user
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:270]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 4]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.unknown
+             
+        # # OK
+        # maxEvaluations='maxEvaluations'                 # Exceeded maximum number of problem evaluations
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:282]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 5]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
-                SolverStatus.aborted
+                SolverStatus.ok
             results.solver.termination_condition = \
                 TerminationCondition.maxEvaluations
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.stoppedByLimit
+                    
+        # # OK
+        # maxEvaluations='maxEvaluations'                 # Exceeded maximum number of problem evaluations
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:295]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 6]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:307]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 7]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
-                SolverStatus.aborted
+                SolverStatus.ok
             results.solver.termination_condition = \
                 TerminationCondition.maxEvaluations
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.stoppedByLimit
+                    
+        # # OK
+        # maxTimeLimit='maxTimeLimit'                     # Exceeded maximum time limited allowed by user
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:319]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 8]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
-                SolverStatus.aborted
+                SolverStatus.ok
             results.solver.termination_condition = \
                 TerminationCondition.maxTimeLimit
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.stoppedByLimit
+                    
+        # # OK
+        # other='other'                                   # Other, uncategorized normal termination
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:332]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 9]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
-                SolverStatus.aborted
+                SolverStatus.ok
             results.solver.termination_condition = \
                 TerminationCondition.other
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.stoppedByLimit
+                    
+        # # OK
+        # other='other'                                   # Other, uncategorized normal termination
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:344]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 10]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
             results.solver.status = \
-                SolverStatus.aborted
+                SolverStatus.ok
             results.solver.termination_condition = \
                 TerminationCondition.other
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.stoppedByLimit
+                    
+        # # OK
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:355]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 11]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
+        # other='other'                                   # Other, uncategorized normal termination
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:368]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 12]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
-                SolverStatus.aborted
+                SolverStatus.ok
             results.solver.termination_condition = \
                 TerminationCondition.other
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.stoppedByLimit
+                    
+        # # OK
+        # optimal='optimal'                               # Found an optimal solution
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:380]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 13]
```diff
@@ -288,6 +387,10 @@ def _postsolve(self):
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.optimal
+                    
+        # # WARNING
+        # infeasible='infeasible'                         # Demonstrated that the problem is infeasible
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:392]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 14]
```diff
@@ -296,6 +399,10 @@ def _postsolve(self):
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.infeasible
+                    
+        # # WARNING
+        # unbounded='unbounded'                           # Demonstrated that problem is unbounded
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:404]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 15]
```diff
@@ -304,6 +411,10 @@ def _postsolve(self):
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.unbounded
+                    
+        # # WARNING
+        # infeasibleOrUnbounded='infeasibleOrUnbounded'   # Problem is either infeasible or unbounded
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:416]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 16]
```diff
@@ -312,6 +423,10 @@ def _postsolve(self):
             if len(results.solution) > 0:
                 results.solution(0).status = \
                     SolutionStatus.unsure
+                    
+        # # UNKNOWN
+        # unknown='unknown'                               # An unitialized value
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:428]: I appreciate these comments being inserted, but can they please be cleaned up a bit? I.e., one comment per line?

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 17]
```diff
@@ -324,3 +439,66 @@ def _postsolve(self):
                     SolutionStatus.unknown
 
         return results
+    
+    @staticmethod
+    def read_scip_log(filename: str):
```
Comment (mrmundt) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:444]: We don't generally use type hints across Pyomo. I would recommend removing this and instead using more detail docstrings (as is standard elsewhere).

**Review by mrmundt — CHANGES_REQUESTED** [review]
General comment: Are there tests that check against these changes? If not, please add some.

**Comment by mrmundt** [comment]
@pmlpm1986 - For your knowledge, the failures are:

```
pyomo.solvers.tests.checks.test_no_solution_behavior.Test_LP_infeasible1_kernel.test_scip_nl

filename = '/tmp/tmpm8k1k99u_scipampl.log'

    @staticmethod
    def read_scip_log(filename: str):
        """Read the SCIP log file and get relevant information from it."""
    
        from collections import deque
    
        with open(filename) as f:
    
            scip_lines = list(deque(f, 7))
            scip_lines.pop()
    
        # get data
    
        colon_position = 19 # or scip_lines[0].index(':')
    
        solver_status = scip_lines[0][colon_position+2:scip_lines[0].index('\n')]
    
        solving_time = float(
            scip_lines[1][colon_position+2:scip_lines[1].index('\n')]
            )
    
        try:
            solving_nodes = int(
>               scip_lines[2][colon_position+2:scip_lines[2].index('(')]
                )
E               ValueError: substring not found

/home/jenkins/workspace/python/pyomo/pyomo/solvers/plugins/solvers/SCIPAMPL.py:466: ValueError

During handling of the above exception, another exception occurred:

self = <pyomo.solvers.tests.checks.test_no_solution_behavior.Test_LP_infeasible1_kernel testMethod=test_scip_nl>

    def return_test(self):
>       return failed_solve_test(self)

/home/jenkins/workspace/python/pyomo/pyomo/solvers/tests/checks/test_no_solution_behavior.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/jenkins/workspace/python/pyomo/pyomo/solvers/tests/checks/test_no_solution_behavior.py:63: in failed_solve_test
    opt, results = model_class.solve(
/home/jenkins/workspace/python/pyomo/pyomo/solvers/tests/models/base.py:128: in solve
    results = opt.solve(
/home/jenkins/workspace/python/pyomo/pyomo/opt/base/solvers.py:602: in solve
    result = self._postsolve()
/home/jenkins/workspace/python/pyomo/pyomo/solvers/plugins/solvers/SCIPAMPL.py:228: in _postsolve
    log_dict = self.read_scip_log(self._log_file)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

filename = '/tmp/tmpm8k1k99u_scipampl.log'

    @staticmethod
    def read_scip_log(filename: str):
        """Read the SCIP log file and get relevant information from it."""
    
        from collections import deque
    
        with open(filename) as f:
    
            scip_lines = list(deque(f, 7))
            scip_lines.pop()
    
        # get data
    
        colon_position = 19 # or scip_lines[0].index(':')
    
        solver_status = scip_lines[0][colon_position+2:scip_lines[0].index('\n')]
    
        solving_time = float(
            scip_lines[1][colon_position+2:scip_lines[1].index('\n')]
            )
    
        try:
            solving_nodes = int(
                scip_lines[2][colon_position+2:scip_lines[2].index('(')]
                )
        except ValueError:
    
>           solving_nodes = int(
                scip_lines[2][colon_position+2:scip_lines[2].index('\n')]
                )
E           ValueError: invalid literal for int() with base 10: 'infinite'

/home/jenkins/workspace/python/pyomo/pyomo/solvers/plugins/solvers/SCIPAMPL.py:470: ValueError

```

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 18]
```diff
@@ -216,70 +277,108 @@ def _postsolve(self):
...
                     SolutionStatus.unknown
+             
+        # # OK
+        # maxEvaluations='maxEvaluations'                 # Exceeded maximum number of problem evaluations
+                    
         elif results.solver.message == "node limit reached":
+        
             results.solver.status = \
-                SolverStatus.aborted
+                SolverStatus.ok
```
[the remark this answers is not in the document]
Comment (pmlpm1986) [reply pyomo/solvers/plugins/solvers/SCIPAMPL.py:287]: I just sought to make a few of them reflect the interpretation they have in:
https://github.com/Pyomo/pyomo/blob/main/pyomo/opt/results/solver.py
According to this file, each termination condition implies a given solver status. As things are, these implications are not observed when using SCIP.
Without these changes, I would get warnings saying that a solution was loaded despite the solver status being 'aborted'. I have not tested backwards compatibility nor with other platforms (I use GNU+linux) but I did create a test file.

**Comment by blnicho** [comment]
@pmlpm1986 any update on this PR?

**Comment by blnicho** [comment]
@pmlpm1986 the next Pyomo release is scheduled for mid-August so if this is something you would like included in the next release we only have a few weeks to resolve the remaining testing issues.

**Comment by pmlpm1986** [comment]
I will revisit this PR soon. I have identified one of the errors but I know very little about 'Jenkins' and the test scripts included with pyomo. I tried running a few but they failed, even without my code inserted, which is why I just went with my own test script.

**Comment by blnicho** [comment]
@pmlpm1986 any update on when you'll get back to this PR? We're waiting to review it until the tests are passing but if you have questions about our testing infrastructure let us know.

**Comment by pmlpm1986** [comment]
I tried running the SCIP-specific tests locally but was unable to, even without my changes inserted. I can run the tests, but I run into errors  (specifically with test_scip.py; test_scip_version.py seems to issue a warning but does not fail), and I have no clue about how to address them. Guidance would be most welcome.
I can also say I tested the changes I introduced (via my own test script) with Python 3.8 and 3.9 and SCIP 8.0.1.

**Review by mrmundt — APPROVED** [review]

**Review by blnicho — APPROVED** [review]

**Pull request merged**